ci: Retry image building once on failure

maflcko opened the pull request:
This should fix https://github.com/bitcoin/bitcoin/issues/33640.

It also contains a few refactor cleanups, which are explained in the corresponding commits.

Commit:
ci: Use os.environ[key] access when value must be set

The other code in this file is using this pattern to throw when a key is
unset, instead of silently returning a None when using os.getenv(key)
with no default value specified.

So use the pattern here as well. As the env vars are always set, this
should be a refactor that does not change the behavior.

Commit:
ci: Allow overwriting check option in run() helper

The bool is forced to check=True, but some commands may want to
explicitly use kwargs to disable it.

This refactor is needed for the next commit.

Commit:
ci: Retry image building once on failure

The build scripts inside the image retry after a failure. However, there
may be some rare network failures inside the container engine. For
example, when pulling the underlying base image, or when pulling the
docker cache.

Thus, retry after a failure once, which should hopefully fix
https://github.com/bitcoin/bitcoin/issues/33640.

Commit:
ci: Export the container id in python script

This refactor does not change behavior, but it has a few benefits:

* The shellcheck SC2086 warning is disabled for the whole command, but
  is only needed for the CI_CONTAINER_CAP env var. So in Python, only
  pass this one env var to shlex.split() for proper word splitting.
* Future logic improvements can be implemented in Python.

The comments are moved, which can be checked via the git options:
--color-moved=dimmed-zebra --color-moved-ws=ignore-all-space

kevkevinpal commented on ci/test/02_run_container.py:
```diff
@@ -71,6 +71,79 @@ def main():
+            # Prune images used by --external containers (e.g. build containers) when
+            # using podman.
+            print("Prune all dangling images")
+            run(["podman", "image", "prune", "--force", "--external"], check=False)
```
Why add `check=False` for these commands when previously there was no `|| true` appended to the original statements

kevkevinpal commented on ci/test/02_run_container.py:
```diff
@@ -64,7 +66,83 @@ def main():
+            f"--mount={CI_DEPENDS_MOUNT}",
+            f"--mount={CI_DEPENDS_SOURCES_MOUNT}",
+            f"--mount={CI_PREVIOUS_RELEASES_MOUNT}",
+            *CI_BUILD_MOUNT,
```
yea maybe something like this? Let me know if this looks good or if there are any issues you see with it

https://github.com/kevkevinpal/bitcoin/pull/184/commits/558c116d3fa49e3a103d70e0c18c873c1d534ae8

kevkevinpal commented on ci/test/02_run_container.py:
```diff
@@ -71,6 +71,79 @@ def main():
+            CI_DEPENDS_MOUNT = f"type=bind,src={os.environ['DEPENDS_DIR']}/built,dst={os.environ['DEPENDS_DIR']}/built"
+            CI_DEPENDS_SOURCES_MOUNT = f"type=bind,src={os.environ['DEPENDS_DIR']}/sources,dst={os.environ['DEPENDS_DIR']}/sources"
+            CI_PREVIOUS_RELEASES_MOUNT = f"type=bind,src={os.environ['PREVIOUS_RELEASES_DIR']},dst={os.environ['PREVIOUS_RELEASES_DIR']}"
+            CI_BUILD_MOUNT = [f"--mount=type=bind,src={os.environ['BASE_BUILD_DIR']},dst={os.environ['BASE_BUILD_DIR']}"]
```
yea maybe something like this? Let me know if this looks good or if there are any issues you see with it

https://github.com/kevkevinpal/bitcoin/pull/184/commits/558c116d3fa49e3a103d70e0c18c873c1d534ae8

l0rinc commented on ci/test/02_run_container.py:
```diff
@@ -71,6 +71,79 @@ def main():
+        run(["docker", "image", "prune", "--force", "--filter", f"label={CI_IMAGE_LABEL}"])
+
+        cmd_run = ["docker", "run", "--rm", "--interactive", "--detach", "--tty"]
+        cmd_run += [
```
nit to avoid modifications:
```suggestion
        cmd_run = [
              "docker", "run", "--rm", "--interactive", "--detach", "--tty"
```

maflcko replied:
I don't think this works with auto-formatters

l0rinc commented on ci/test/02_run_container.sh:
```diff
@@ -5,76 +5,10 @@
-                  --platform="${CI_IMAGE_PLATFORM}" \
-                  "$CONTAINER_NAME")
-  export CI_CONTAINER_ID
   export CI_EXEC_CMD_PREFIX="docker exec ${CI_CONTAINER_ID}"
```
does this still work correctly now that `CI_CONTAINER_ID` is not exported?

maflcko replied:
Yes, it should be correct. My understanding is that `export CI_CONTAINER_ID` is roughly equal to `os.environ["CI_CONTAINER_ID"] = container_id` in Python.

You can also see in the CI logs that it works. https://github.com/bitcoin/bitcoin/actions/runs/18776395325/job/53571955105?pr=33677#step:9:247

Commit:
ci: Document why IN_GETOPT_BIN env var is needed on macOS

This was added in commit b705bade44973e61655d5f847f49d97fb5bb8393, but I
keep forgetting the background that this is needed for the retry Bash
script. So document it.

l0rinc commented:
Code review ACK fa586e86c4aa20c07f37488c13b2b797b35fdbd9

The small focused commits with detailed messages made the review very easy.
I left a few suggestions for cases that could be made more pythonic (i.e. partial bash migration), but I don't mind keeping it as is, seems correct to me.

maflcko commented on ci/test/02_run_container.py:
```diff
@@ -64,7 +66,83 @@ def main():
+            cmd_run,
+            stdout=subprocess.PIPE,
+            text=True,
+            encoding="ascii",
```
I actually don't know if it is needed. The existing code and `test/lint/lint-python-utf8-encoding.py` give different vibes. Not sure what to do, but I can remove it in a follow-up or the next push.

maflcko commented on ci/test/02_run_container.py:
```diff
@@ -65,7 +66,10 @@ def main():
 
         print(f"Building {os.environ['CONTAINER_NAME']} image tag to run in")
-        run(cmd_build)
+        if run(cmd_build, check=False).returncode != 0:
```
Yes, but `run` does not throw. You can find the function above and it behaves roughly equal to `set -o errexit` in Bash (exit immediately on failure).
